stablehlo/reference/Element.cpp

```diff
@@ -720,6 +720,57 @@ Element rsqrt(const Element &el) {
       [](std::complex<double> e) { return 1.0 / std::sqrt(e); });
 }
 
+Element sign(const Element &el) {
+  Type type = el.getType();
+
+  if (isSupportedIntegerType(type)) {
+    auto intEl = el.getIntegerValue();
+    if (intEl.isZero()) return Element(type, (int64_t)0);
+    if (intEl.isNegative())
+      return Element(type, (int64_t)-1);
+    else
+      return Element(type, (int64_t)1);
+  }
+
+  if (isSupportedFloatType(type)) {
+    auto elVal = el.getFloatValue();
+
+    if (elVal.isNaN()) return Element(el);
+    if (elVal.isNegZero()) return Element(type, (double)-0.0);
+    if (elVal.isPosZero()) return Element(type, (double)0.0);
+    if (elVal.isNegative()) return Element(type, (double)-1.0);
+    if (!elVal.isNegative()) return Element(type, (double)1.0);
+  }
+
+  if (isSupportedComplexType(type)) {
+    auto elVal = el.getComplexValue();
+    const llvm::fltSemantics &elSemantics = elVal.real().getSemantics();
+
+    if (elVal.real().isNaN() || elVal.imag().isNaN())
+      return Element(type,
+                     std::complex<APFloat>(APFloat::getNaN(elSemantics),
+                                           APFloat::getZero(elSemantics)));
+
+    if (elVal.real().isZero() && elVal.imag().isZero())
+      return Element(type,
+                     std::complex<APFloat>(APFloat::getZero(elSemantics),
+                                           APFloat::getZero(elSemantics)));
+
+    auto absElValDouble = abs(el).getFloatValue().convertToDouble();
+    bool roundingErr;
+
+    APFloat resultReal(elVal.real().convertToDouble() / absElValDouble);
+    resultReal.convert(elSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    APFloat resultImag(elVal.imag().convertToDouble() / absElValDouble);
+    resultImag.convert(elSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+
+    return Element(type, std::complex<APFloat>(resultReal, resultImag));
```

Comment:
Can we reformulate the complex case in terms of already existing functions, i.e. `el / abs(el)`? (While still keeping the special logic that handles NaNs).

Reply:
Hello @burmako , I first tried making the code to be as simple as `el / abs(el)`.

But the spec for `divide` operation says:
* (C1) `lhs`, `rhs` and `result` have the same type.

The spec of `abs` says:
* (C2)  `operand` and `result` have the same element type, except when the element type of the `operand` is complex type, in which case the element type of the `result` is the element type of the complex type (e.g. the element type of the `result` is `f64` for operand type `complex<f64>`).

So for a given floating-point type `T` and number of elements `N`:

`tensor<Nxcomplex<T>> / abs(tensor<Nxcomplex<T>>)` becomes `tensor<Nxcomplex<T>> / tensor<NxT>`

and the latter is not supported. So that's why I wrote the code like this.

If I have assumed or understood anything incorrectly please do correct me and I shall make the changes accordingly.

Reply:
You're right - this also indicates a gap in the spec, which says `divide(x, abs(x))`. Let me think how to best fix this in the spec. Will follow up in a few hours after my meetings.

Reply:
Sure, will do.

Also, is there a plan to include `complex / float`, `complex / int`, `float / int` in the spec ?

Like wise for all other ops `+`, `-`, and `*`.

Reply:
I think that the idiomatic way of going about this in HLO is to use `convert`: https://github.com/openxla/stablehlo/blob/main/docs/spec.md#convert, so we probably should do that as well.

Reply:
#1349 implements it. I should be able to leverage that when it is merged.

Reply:
I think we don't even have to wait for convert to be merged, since there's already a compact way of expressing this: `return el / Element(type, abs(el).getFloatValue().convertToDouble())`.